### PR TITLE
MIR peephole optimize {Ne, Eq}(_1, false) into _1

### PR DESCRIPTION
**Pull request by simonvandel**
Add peephole optimization that simplifies Ne(_1, false) and Ne(false, _1) into _1. Similarly handles Eq(_1, true) and Eq(true, _1).

This was observed emitted from the MatchBranchSimplification pass.

### CODE REVIEW

File: src/librustc_mir/transform/instcombine.rs
Comment (pickfire) [inline src/librustc_mir/transform/instcombine.rs:133]: Why not just show the representation? The next feels hard to comprehend.
```suggestion
            // (false, _place)
```

Comment (simonvandel) [reply src/librustc_mir/transform/instcombine.rs]: Good point, fixed

**Review by pickfire — APPROVED** [review]
I don't know much about peephole optimization but looks good to me.

**Comment by pickfire** [comment]
Would this affect compile time? Should we run a timer?

**Comment by simonvandel** [comment]
My intuition says that this shouldn't be that expensive but let's try perf bot to be sure?

**Comment by pickfire** [comment]
@jyn514 Can you please help to run perf bot?

**Review by davidtwco — APPROVED** [review]
LGTM, r=me after perf

**Comment by simonvandel** [comment]
@davidtwco rebased on master to fix merge conflicts. It should be ready for a new bors try

**Comment by davidtwco** [comment]
I'm not sure what conclusions can be drawn from the perf results - going to let someone else take a look before merging. 

r? @wesleywiser 

> @davidtwco rebased on master to fix merge conflicts. It should be ready for a new bors try

I don't think we'll need another perf run, so another try build isn't necessary.

**Comment by wesleywiser** [comment]
This looks good to me but I think for consistency we should do the same thing for `Eq(_, true) | Eq(true, _)`. @simonvandel if you don't want to do that in the PR, let me know so we can open an issue instead and someone else can work on it.

**Comment by simonvandel** [comment]
I can extend the PR to also handle the Eq case today, sure.

**Comment by simonvandel** [comment]
@wesleywiser added a new commit so that Eq is handled similarly

**Comment by wesleywiser** [comment]
We usually do not rollup MIR optimizations even if they don't seem to have a performance impact because it would be best for bisection purposes if this does not end up in a rollup. Therefore

@bors r+ rollup=never

**Comment by pickfire** [comment]
@wesleywiser Do we usually not run timer for this?

**Comment by wesleywiser** [comment]
@pickfire I think since just the `Ne` optimization had little effect, both the `Ne` and `Eq` ones will probably not have a huge effect either. perf.rlo will run on the merge commit so we will be able to see exactly what the impact is.

**Comment by simonvandel** [comment]
So I don't know why exactly the test failed:
```
-	  fn opt(_1: std::option::Option<()>) -> bool {
+	  fn opt(_1: Option<()>) -> bool {
```
I rebased on master locally, rebless and that did change my test. So i have now force pushed that. @wesleywiser can you start the bors dance again?

**Pull request merged**
